keyserver/server.go

```diff
@@ -0,0 +1,177 @@
+package keyserver
+
+import (
+	"crypto/tls"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/BurntSushi/toml"
+	"github.com/coniks-sys/coniks-go/crypto/sign"
+	"github.com/coniks-sys/coniks-go/protocol"
+	"github.com/coniks-sys/coniks-go/storage/kv"
+	"github.com/coniks-sys/coniks-go/storage/kv/leveldbkv"
+)
+
+type ServerConfig struct {
+	SigningKeyPath       string  `toml:"signing_key"`
+	DatabasePath         string  `toml:"database"`
+	RegistrationAddress  string  `toml:"reg_address"`
+	Address              string  `toml:"address"` // address:port
+	LoadedHistoryLength  uint64  `toml:"loaded_history_length"`
+	RegistrationCapacity uint64  `toml:"registration_capacity"`
+	PoliciesPath         string  `toml:"policies"`
+	TLS                  TLSInfo `toml:"tls"`
+}
+
+type TLSInfo struct {
+	TLSCertPath string `toml:"cert"`
+	TLSKeyPath  string `toml:"key"`
+}
+
+type ConiksServer struct {
+	sync.Mutex
+	directory *protocol.ConiksDirectory
+
+	stop     chan struct{}
+	waitStop sync.WaitGroup
+
+	db kv.DB
+
+	policies         *ServerPolicies
+	policiesFilePath string // should be final
+	policiesMutex    sync.Mutex
+	reloadChan       chan os.Signal
+	epochTimer       *time.Timer
+}
+
+func LoadServerConfig(path string) *ServerConfig {
+	var conf ServerConfig
+	if _, err := toml.DecodeFile(path, &conf); err != nil {
+		log.Fatalf("Failed to load config: %v", err)
+		return nil
+	}
+	return &conf
+}
+
+func New(conf *ServerConfig) *ConiksServer {
+	// load signing key
+	skBytes, err := ioutil.ReadFile(conf.SigningKeyPath)
+	if err != nil {
+		log.Fatalf("Cannot read signing key: %v", err)
+		return nil
+	}
+	if len(skBytes) != sign.PrivateKeySize {
+		log.Fatalf("Signing key must be 64 bytes (got %d)", len(skBytes))
+		return nil
+	}
+
+	sk := make([]byte, sign.PrivateKeySize)
+	copy(sk, skBytes[:sign.PrivateKeySize])
+
+	// open db
+	kvdb := leveldbkv.OpenDB(conf.DatabasePath)
+
+	// read server policies
+	policies, vrfKey, err := readPolicies(conf.PoliciesPath)
+	if err != nil {
+		log.Fatalf("Cannot read policies config: %v", err)
+		return nil
+	}
+
+	// create server instance
+	server := new(ConiksServer)
+	server.stop = make(chan struct{})
+	server.db = kvdb
+	server.policies = policies
+	server.policiesFilePath = conf.PoliciesPath
+	server.reloadChan = make(chan os.Signal, 1)
+	signal.Notify(server.reloadChan, syscall.SIGUSR2)
+	server.epochTimer = time.NewTimer(time.Duration(policies.EpochDeadline) * time.Second)
+	server.directory = protocol.InitDirectory(
+		policies.EpochDeadline, vrfKey,
+		sk, conf.LoadedHistoryLength,
+		true, conf.RegistrationCapacity)
+
+	return server
+}
+
+func (server *ConiksServer) RunWithConfig(conf *ServerConfig) {
```

Comment:
As opposed to just `Run` (which doesn't exist)?  You've passed in `conf` at construction time.  Should you be holding a pointer to it?

The question here is, is the `conf` expected to differ between construction and use?